### DISCUSSION TITLE
Self host WASM with `dotlottie-wc`?

### DISCUSSION THREAD

**Discussion by AlanBreck** [LottieFiles/dotlottie-web#198] [Q&A] [answered]
I've noticed that the WASM file is served from https://www.jsdelivr.com/ when using `@lottiefiles/dotlottie-wc`, and it's not downloaded with the package. Is there any way of self-hosting the WASM?

I also see that there's a static method `setWasmUrl`, but I'm not sure how to properly obtain the WASM in order to self host.

**Comment by miszo** [comment]
I have the same question about the `@lottiefiles/dotlottie-web` and `@lottiefiles/dotlottie-react`.
Tried to do the following, but it doesn't seem to work

```ts
import { DotLottie } from '@lottiefiles/dotlottie-web';
import lottieWasmUrl from '@lottiefiles/dotlottie-web/dist/dotlottie-player.wasm'; // this is loaded as a url by the bundler

export function setLottieWasmUrl() {
  DotLottie.setWasmUrl(lottieWasmUrl);
}
```
I was trying to use it on the component level and when bootstrapping the application but every time I get the WASM file loaded from the jsdelivr.com.

I'm using the following versions:
* `@lottiefiles/dotlottie-web`: 0.20.1 (didn't include it in my package.json, used as a dependency from the react package)
* `@lottiefiles/dotlottie-react`: 0.5.5

![CleanShot 2024-05-17 at 18 47 21@2x](https://github.com/LottieFiles/dotlottie-web/assets/3455863/94a0e780-17bf-43bb-a1e8-69143483c4e2)

**Comment by theashraf** [comment]
We currently do not have a self-contained bundle that embeds the WASM file within the JavaScript, making it self-contained. The downside of this approach is the increased JavaScript bundle size, which could reach approximately 1 MB. To keep the bundle size smaller, we preferred shipping the WASM file independently and asynchronously loading it from the CDN (jsdelivr). As a fallback, we also added unpkg CDN.

If you want to self-host the WASM file, you can use `dotlottie-web` directly and load the `dotlottie-player` with a URL loader.

### Using as a React Component

```tsx
import { DotLottie } from "@lottiefiles/dotlottie-web";
import wasmUrl from "@lottiefiles/dotlottie-web/dist/dotlottie-player.wasm?url";
import { useRef, useEffect } from "react";

// Set the WASM URL
DotLottie.setWasmUrl(wasmUrl);

function DotLottiePlayer() {
  const ref = useRef<HTMLCanvasElement>(null);

  useEffect(() => {
    new DotLottie({
      canvas: ref.current!,
      src: "https://lottie.host/242ebeec-6277-4937-afa6-3a1f9f54f0c0/UFK246mvwd.lottie",
      autoplay: true,
      loop: true,
    });
  }, []);

  return (
    <canvas
      ref={ref}
      style={{
        width: 300,
        height: 300,
      }}
    />
  );
}

export default DotLottiePlayer;
```

### Using as a Web Component

```jsx
import { DotLottie } from "@lottiefiles/dotlottie-web";
import wasmUrl from "@lottiefiles/dotlottie-web/dist/dotlottie-player.wasm?url";

// Set the WASM URL
DotLottie.setWasmUrl(wasmUrl);

class DotLottiePlayer extends HTMLElement {
  constructor() {
    super();
    this.attachShadow({ mode: "open" });
    this.canvas = document.createElement("canvas");
    this.canvas.style.width = "300px";
    this.canvas.style.height = "300px";
    this.shadowRoot.appendChild(this.canvas);
  }

  static get observedAttributes() {
    return ["src", "loop", "autoplay"];
  }

  attributeChangedCallback(name, oldValue, newValue) {
    if (name === "src" || name === "loop" || name === "autoplay") {
      this.loadAnimation();
    }
  }

  connectedCallback() {
    this.loadAnimation();
  }

  loadAnimation() {
    const src = this.getAttribute("src");
    const loop = this.getAttribute("loop") === "true";
    const autoplay = this.getAttribute("autoplay") === "true";

    new DotLottie({
      canvas: this.canvas,
      src: src,
      autoplay: autoplay,
      loop: loop,
    });
  }
}

customElements.define("dotlottie-player", DotLottiePlayer);
```

**Reply by miszo** [reply]
Thanks for the reply @theashraf. Is it somehow possible to self-host the WASM file and using your React wrapper? Maybe you could export `DotLottie` from the `@lottiefiles/dotlottie-react` as well? Or perhaps some wrapper for the `DotLottie.setWasmUrl`?

**Reply by AlanBreck** [reply]
Thanks, @theashraf! Is there any possibility in future development that this could be configured via a prop for the various packages (React, WC, Svelte, etc)?

**Reply by theashraf** [reply]
@miszo @AlanBreck Sure, I think we can export a util function called `setWasmUrl` from all the dotlottie-web wrappers, I'll keep you updated

**Reply by theashraf** [reply]
@AlanBreck @miszo I've raised a PR for this feature: https://github.com/LottieFiles/dotlottie-web/pull/204. Please have a look and let me know if it works for you 👀

**Reply by theashraf** [reply, marked as answer]
@AlanBreck @miszo This update has landed in `dotlottie-react@0.6.0` and `dotlottie-wc@0.2.0`, please have a look and let me know

**Reply by adelegard** [reply]
@theashraf this is awesome. However when using a local url/file I'm seeing both the local file as well as the backup be loaded. I'm on version `"@lottiefiles/dotlottie-react": "0.6.3",`

```typescript
import { DotLottie, DotLottieReact, setWasmUrl } from '@lottiefiles/dotlottie-react'
setWasmUrl('/lottie/dotlottie-player.wasm')
```

<img width="850" alt="Screenshot 2024-06-07 at 10 44 58 AM" src="https://github.com/LottieFiles/dotlottie-web/assets/209605/79be9bdc-1d42-4319-b3fe-36150c4abac5">
<img width="846" alt="Screenshot 2024-06-07 at 10 44 41 AM" src="https://github.com/LottieFiles/dotlottie-web/assets/209605/55a7264b-391e-4bec-a02d-a947bbd5e924">
<img width="848" alt="Screenshot 2024-06-07 at 10 44 34 AM" src="https://github.com/LottieFiles/dotlottie-web/assets/209605/67a94249-97df-4029-bc21-712c619a9097">

**Reply by AlanBreck** [reply]
~@adelegard, I believe that means that it can't find the WASM at the path you've provided. I'd recommend confirming that `/lottie/dotlottie-player.wasm` is indeed the correct path by accessing via a `fetch`, or something.~

NM, I didn't look at your screenshots closely enough. Sorry for noise.

**Reply by theashraf** [reply]
@adelegard The backup WASM URL will be used only if loading the WASM file from the provided URL in the `setWasmUrl` method fails. It would be very helpful if you could provide a CodeSandbox with this issue replicated. I'd be happy to take a look and assist further

**Reply by pclarke-coursera** [reply]
> We currently do not have a self-contained bundle that embeds the WASM file within the JavaScript, making it self-contained. The downside of this approach is the increased JavaScript bundle size, which could reach approximately 1 MB. To keep the bundle size smaller, we preferred shipping the WASM file independently and asynchronously loading it from the CDN (jsdelivr). As a fallback, we also added unpkg CDN.
> 
> If you want to self-host the WASM file, you can use `dotlottie-web` directly and load the `dotlottie-player` with a URL loader.
> 
> ### Using as a React Component
> ```tsx
> import { DotLottie } from "@lottiefiles/dotlottie-web";
> import wasmUrl from "@lottiefiles/dotlottie-web/dist/dotlottie-player.wasm?url";
> import { useRef, useEffect } from "react";
> 
> // Set the WASM URL
> DotLottie.setWasmUrl(wasmUrl);
> 
> function DotLottiePlayer() {
>   const ref = useRef<HTMLCanvasElement>(null);
> 
>   useEffect(() => {
>     new DotLottie({
>       canvas: ref.current!,
>       src: "https://lottie.host/242ebeec-6277-4937-afa6-3a1f9f54f0c0/UFK246mvwd.lottie",
>       autoplay: true,
>       loop: true,
>     });
>   }, []);
> 
>   return (
>     <canvas
>       ref={ref}
>       style={{
>         width: 300,
>         height: 300,
>       }}
>     />
>   );
> }
> 
> export default DotLottiePlayer;
> ```


@theashraf Is there typescript support for `import wasmUrl from "@lottiefiles/dotlottie-web/dist/dotlottie-player.wasm?url";`?

**Reply by pclarke-coursera** [reply]
Ah looks like the `?url` piece might be specific to vite
